### PR TITLE
add campsite database and map

### LINKED ISSUE #5358: Create Campsite Database, Interface, Connect to Google API
We need a campsite database to simplify management of campsites and to auto-populate a map showing the locations of campsites to allow users to find a campsite close to them.
### Database Structure

| Field Name | Type | Notes |
| --- | --- | --- |
| id | ObjectId | Unique Campsite ID |
| city | string | City Name |
| country | string | Country Name - Populate from Dropdown |
| subdivision | string | State/Province, etc |
| coordinates | object | Lat/Long object |
| google_id | string | unique google id for map element |
| url | string | Social Media URL (IE: Facebook Link) |
| createdBy | ObjectId | FCC User who created the record |
| createdOn | date | Campsite Record Creation date |
| display | boolean | Display/Don't Display (for Moderation) |
### Interface

**Route**: /campsites
- Default View
  Simple single page display with Embedded Google Map.  Below the map is a hierarchical, alphabetic display of all campsites, same as on the Wiki page.
- Add Campsite Button 
  Pops up a modal with the appropriate fields.
  Do Google API lookup on submit to validate location of city and fill in sub-division?
  On submit send an e-mail to moderators?
- Moderation
  Table on default view shows "unapproved" campsites
  Click "Approve" to toggle display state, posts to Google API
  Click "Delete" to remove campsite
  Click "Edit" to change location, etc.
### Google API

Upon approval, coordinates are sent to the Google API to update the map.

### PR DESCRIPTION
**Pull request by geoff616**
### Summary:

This pull request implements the features described in [issue 5358](https://github.com/FreeCodeCamp/FreeCodeCamp/issues/5358) to help simplify the management of campsites.
### User Stories:
- Users can see a list of all campsites and a map with a pin on each location
- Users can add new campsites after logging in
- Campsite admins can see how many new campsites have been submitted
- Campsite admins can review pending campsites before approving 
- Campsite admins can edit existing campsites
### Implementation:

`Campsites` is a React app modeled after the `Jobs` and `Hikes` apps and includes:
- Adding five new routes:
  - `/campsites` - campsite list and map
  - `/campsites/new` - add a new campsite
  - `/campsites/admin` - campsite admin homepage
  - `/campsites/admin/pending` - campsite admin can review pending campsites
  - `/campsites/admin/edit/:id` - campsite admin can edit existing campsites
- Adding one model and updating another:
  - added `campsite` - includes all information about a campsite
  - updated `user` - added a boolean `isCampsiteModerator`
- Adding one service:
  - `campsites` - with `read`, `create`, and `update` methods
- Adding two dependencies to `package.json`:
  - [`react-google-maps`](https://github.com/tomchentw/react-google-maps) - This library takes care of asynchronously loading and rendering the maps used to display and search for campsites. 
  - [`react-select`](https://github.com/JedWatson/react-select) - This library provides inputs fields with dropdowns, and is used by moderators when updating campsite locations.
#### Security:
- **User Input:** User input is sanitized like in the `Jobs` app. 
- **Campsite Model API Routes:** These are currently public and used by the bulk upload script (described in Migrating Existing Campsites), but could be closed to prevent working around moderation.
- **allowExtendedOperators:** This is set to true in the `campsites` model to allow update operations, but could be scoped to the specific update requests coming from the `campsites` service if this is a concern. 
- **Trusted Script Sources:** Added `maps.googleapis.com` to trusted script sources
#### Migrating existing campsites:

Campsites that are listed on the [Wiki](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/List-of-Free-Code-Camp-city-based-Campsites) can be migrated to the app with a [bulk upload script](https://github.com/geoff616/parsingCampsites).

For 99%+ of campsites, it was possible to parse the information in the wiki into a request for the Google Maps API. During testing, 4 campsites were not included in the migration. The details can be seen in [28] [here](https://github.com/geoff616/parsingCampsites/blob/master/generateJSON.ipynb) and the latest campsites should be tested before running this in production to determine if this number has increased, and what to do with these cases. 

This script is meant to be run once, and is not idempotent
#### Creating Campsite Administrators

Creating an admin requires setting the `isCampsiteModerator` property to true for a user.

Running FCC locally, this is achieved by PUTing `{'isCampsiteModerator': true}` to: http://127.0.0.1:3000/api/users/:userID in [the API explorer](http://127.0.0.1:3000/explorer/#!/user/user_prototype_updateAttributes) or another HTTP client.

This route requires authorization in the production app, and it isn't clear if this is the best way to do it there.
#### Performance

Every time the campsite list loads, it is reading 1,000+ records from the DB. This information doesn't change very often, and could benefit from caching if this starts adding significant load to the DB.

Splitting the admin functionality into different containers helped improve rendering performance, especially for `/campsites` and `/campsites/new`, but there is still have some lag when navigating to the admin section from `/campsites`. I haven't been able to identify exactly what is causing it yet, but suspect it might be over-fetching.

I had also played around with adding search functionality to `/campsites` to filter the list and map pins, but wasn't able to get that to performant enough to include yet. 
#### Mobile

Creating campsites on touch screen devices doesn't work with the current implementation of the Google Maps search, and non-admin users on these devices won't see an option to add a campsite. This could be revisited if this becomes an important use case. 

This implementation is also sending 1+ MB of JS back to the client to render the list and map of campsites on `/campsites`. This might be overkill for what is being used by the non-admin users, and could also be revisited if this is a concern. 

closes #5358

### CODE REVIEW

**Comment by BerkeleyTrue** [comment]
We're currently moving to Redux, so this PR will need to refactored once #7214 is merged in.

**Comment by geoff616** [comment]
Sounds good @BerkeleyTrue

**Comment by mikeissokool** [comment]
Sorry everyone I just learned how to get here

**Comment by mikeissokool** [comment]
Ty to everyone who has worked keeping info priv

**Comment by BerkeleyTrue** [comment]
@geoff616 You'll need to rebase instead of merging. And it should probably wait until that pr is merged into staging as there may still be changes.

**Comment by geoff616** [comment]
Thanks @BerkeleyTrue, just rebased against latest staging. Will start moving over to redux.

**Comment by geoff616** [comment]
Moved over to redux, and have most of the functionality working again. Still need to add input validation, implement the **user** related requirements and clean things up. 

Currently a little stuck on [keeping redux and the DB in sync on update operations](https://github.com/geoff616/FreeCodeCamp/commit/6692cde830b1226930197519a998fbb92f3ef7f1#diff-4f6793fc9ab906867640e27658c1055bR34), and haven't found a way to get this working yet.

**Comment by geoff616** [comment]
Have all the functionality working again in redux, and added in the user features.

Wrote [some scripts](https://github.com/geoff616/parsingCampsites) to migrate the MD file with the campsites into the DB. After uploading the 1,100+ campsites, the responsiveness of the app dropped significantly, and I'm looking into ways of making it less janky with that much data. 

Also working through a bug, where edits to existing campsites are not re-rendering correctly. [Asked for some help on SO](http://stackoverflow.com/questions/35965863/keying-react-lists-with-dynamic-properties/35970815), and waiting to see what they think.

Going to fix the above issues and will rewrite the info at the top to reflect the changes to the PR.

File: client/less/campsites.less [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#campsite-buttons {
+  padding-top: 25px;
```
Comment (BerkeleyTrue) [inline client/less/campsites.less]: Try using classes instead of id's for CSS

**Comment by BerkeleyTrue** [comment]
Thanks for the PR. I'm doing a lot of work on moving over the rest of the site to React so I've not been able to give this PR the attention it deserves. From a cursory examination, the code looks solid, but those issue you described will definitely need to be tackled.

**Comment by geoff616** [comment]
Thanks @BerkeleyTrue. Made some progress on resolving those issues, but still working through a few bugs. Should have the PR updated by the end of the week, and happy to revisit when it makes sense.

**Comment by geoff616** [comment]
Updated the information at the top of the page to reflect the latest changes. Let me know if there is anything I can do to help make things more clear or help move this forward.

**Comment by BerkeleyTrue** [comment]
While this is still blocked, I want to keep it open as this is something I want to see merged in. It will have to wait until the rest of the reactification has finished, though.

**Comment by crashuniverse** [comment]
If I were to add my ideas on routes (and it is a trivial opinion) the REST routes should be:

**Current -> I propose:**
/campsites ->  `/campsites`
/campsites/new -> `POST /campsites`
/campsites/admin ->  `/campsites/admin`
/campsites/admin/pending -> `/campsites/?status=pending`
/campsites/admin/edit/:id -> `PUT /campsites/:id`

**Comment by atjonathan** [comment]
@BerkeleyTrue this can probably looked at again once we have finished testing the rectification 👍

**Comment by geoff616** [comment]
Great! Is there a branch I should rebase against, or just wait until everything is on staging?

**Comment by atjonathan** [comment]
@geoff616, everything you need should be on `staging`, so rebase against that :smiley:

**Comment by geoff616** [comment]
Thanks @atjonathan! 

I got started on the rebase and cleaned up the merge conflicts, but ran into a bit of a snag with some of the changes to how the sagas fetch data now. 

Currently getting an error message `500 Invariant Violation: saga should returned an observable but got function` - [full error message](https://gist.github.com/geoff616/9325422e4e97f3201b62bec77b6239e8)

I'll try to dig into the new redux saga stuff this weekend!

**Comment by BerkeleyTrue** [comment]
@geoff616 Are you still able to work on PR?

**Comment by geoff616** [comment]
@BerkeleyTrue I am! Just started a new job, and haven't been able to spend as much time on this as I had wanted. Hoping to have something completed by end of next weekend. 

I've spent some time trying to learn how the new data fetching stuff works, but still don't have a sense of when I'll be able to work with it. 

The clearest path I currently see is hitting the`/api/campsites` routes directly from the client. 

Is that okay, or should I try to work within the current data fetching implementations?

**Comment by hallaathrad** [comment]
Wishing I could speak this weird sanskrit dialect so I could help you move this forward 🤕 
However, if there's anything at all I can help with from the local groups side, please do let me know 👍

File: client/less/main.less [hunk 2]
```diff
@@ -3,6 +3,7 @@
 @import "lib/ionicons/ionicons";
 @import "lib/animate";
 @import "lib/bootstrap/variables";
+@import "../../node_modules/react-select/less/default.less";
```
Comment (shojaala) [inline client/less/main.less]: 6

File: client/less/main.less [hunk 3]
```diff
@@ -1129,10 +1130,11 @@ and (max-width : 400px)  {
...
 
+@import "campsites.less";
 @import "chat.less";
 @import "jobs.less";
 @import 'code-mirror.less';
 @import "challenge.less";
 @import "toastr.less";
 @import "map.less";
-@import "drawers.less";
+@import "drawers.less";
```
Comment (shojaala) [inline client/less/main.less]: 1140

File: common/app/create-reducer.js [hunk 4]
```diff
@@ -4,6 +4,7 @@ import { reducer as formReducer } from 'redux-form';
 import { reducer as app } from './redux';
 import { reducer as toasts } from './toasts/redux';
 import entitiesReducer from './redux/entities-reducer';
+import { reducer as campsitesApp } from './routes/Campsites/redux';
```
Comment (shojaala) [inline common/app/create-reducer.js]: 7

Comment (shojaala) [inline common/app/create-reducer.js]: 7

File: common/app/create-reducer.js [hunk 5]
```diff
@@ -16,6 +17,7 @@ export default function createReducer(sideReducers = {}) {
     entities: entitiesReducer,
     app,
     toasts,
+    campsitesApp,
```
Comment (shojaala) [inline common/app/create-reducer.js]: 20

File: common/app/redux/fetch-user-saga.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ import {
 
 const { fetchUser } = types;
 
+
```
Comment (shojaala) [inline common/app/redux/fetch-user-saga.js]: 13

Comment (shojaala) [inline common/app/redux/fetch-user-saga.js]: 13

Comment (shojaala) [inline common/app/redux/fetch-user-saga.js]: 13

File: client/less/main.less (hunk 2 above)
Comment (shojaala) [inline client/less/main.less]: 6

File: client/less/main.less (hunk 3 above)
Comment (shojaala) [inline client/less/main.less]: 1140

**Comment by evaristoc** [comment]
@geoff616 is this now working using Forum data?

**Comment by QuincyLarson** [comment]
@g1na1011 would you be interested in building this as a stand-alone website - maybe on GoMix? Then we could change NGINX to point freecodecamp.com/local-study-groups to this standalone app. 

This would definitely be better than continuing to use the forum for this. @hallaathrad what are your thoughts on this?

I'm going to close this PR and the original issue. If this is something you're interested in, please approach me :)

**Pull request closed without merging**